JMMServer/Tasks/ContractExtractor.cs

```diff
@@ -0,0 +1,233 @@
+                Projections.Cast(NHibernateUtil.String, Projections.Property(keyCol1)),
+                Projections.Constant(prefix2),
+                Projections.Cast(NHibernateUtil.String, Projections.Property(keyCol2)),
+                Projections.Constant(".json"));
+        }
+
+        private static IProjection CreateEntryNameProjection<T>(string prefix1, Expression<Func<T, object>> keyCol1,
+            string prefix2, Expression<Func<T, object>> keyCol2,
+            string prefix3, Expression<Func<T, object>> keyCol3)
+        {
+            return Projections.SqlFunction("concat", NHibernateUtil.String,
+                Projections.Constant(prefix1),
+                Projections.Cast(NHibernateUtil.String, Projections.Property(keyCol1)),
+                Projections.Constant(prefix2),
+                Projections.Cast(NHibernateUtil.String, Projections.Property(keyCol2)),
+                Projections.Constant(prefix3),
+                Projections.Cast(NHibernateUtil.String, Projections.Property(keyCol3)),
+                Projections.Constant(".json"));
+        }
+
+        /// <summary>
+        /// Gets the contracts as embedded json files within a zip archive.
+        /// </summary>
+        /// <param name="entityType">The type of the entity to dump (can be <see cref="string.Empty"/> or <c>null</c> to dump all).</param>
+        /// <returns>The zip archive <see cref="System.IO.Stream"/>.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="session"/> is <c>null</c>.</exception>
+        /// <exception cref="ArgumentException">The specified <paramref name="entityType"/> is not valid.</exception>
+        public System.IO.Stream GetContractsAsZipStream(string entityType)
+        {
+            using (IStatelessSession session = DatabaseFactory.SessionFactory.OpenStatelessSession())
+            {
+                return GetContractsAsZipStream(session.Wrap(), entityType);
+            }
+        }
+
+        /// <summary>
+        /// Gets the contracts as embedded json files within a zip archive.
+        /// </summary>
+        /// <param name="session">The NHibernate session.</param>
+        /// <param name="entityType">The type of the entity to dump (can be <see cref="string.Empty"/> or <c>null</c> to dump all).</param>
+        /// <returns>The zip archive <see cref="System.IO.Stream"/>.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="session"/> is <c>null</c>.</exception>
+        /// <exception cref="ArgumentException">The specified <paramref name="entityType"/> is not valid.</exception>
+        public System.IO.Stream GetContractsAsZipStream(ISessionWrapper session, string entityType)
+        {
+            if (session == null)
+                throw new ArgumentNullException(nameof(session));
+
+            MemoryStream buffer = new MemoryStream();
+
+            using (ZipArchive zip = new ZipArchive(buffer, ZipArchiveMode.Create, leaveOpen: true))

```

Comment:
and you just explained something else I've wondered about for years....why tar.gz exists